articles/qos_configurability.md

```diff
@@ -0,0 +1,399 @@
+ROS 2 currently provides read only parameters.
+They can be modified when constructing a node by providing overrides (e.g. `--ros-args -p <param_name> <param_value>`), but they cannot be dynamically changed after the node was constructed.
+
+This perfectly matches the goal of the proposal of not making QoS settings reconfigurable during runtime.
+
+Note: parameter overrides can be constructed dynamically and then used to create a node, but it's not possible to modify the parameter after the node construction.
+This allows, for example, to pass parameter overrides when dynamically adding a node to a component container.
+
+### Parameter events
+
+Up to ROS 2 Foxy, read-only parameters generate a parameter event with its initial value when declared.
+If we start declaring parameters for each QoS policies of many entities, the amount of parameter events at startup time will significantly increase, as the declaration of each parameter generates an event.
+
+To avoid this issue, a way to opt-out parameter events when declaring a parameter could be added.
+
+Note: Opting-out generating a parameter event when setting/unsetting a parameter could also be possible, but not required for this.
+
+### Parameter callbacks
+
+If we want the following example to work:
+
+```cpp
+node->create_publisher(
+  "chatter",
+  KeepLast(10),
+  [] (const QoSProfile & qos) -> bool{
+      return qos.is_reliable();
+  });
+```
+
+We could completely ignore parameters [on set callbacks](https://github.com/ros2/rclcpp/blob/3defa8fc9d7410bd833ecd95b305ac94bb9b627a/rclcpp/include/rclcpp/node_interfaces/node_parameters_interface.hpp#L180-L192), the publisher creation code would do something like:
+
+```cpp
+rcl_interfaces::msg::ParameterDescription policy_x_description{};
+policy_x_description.read_only = true;
+policy_x_description.description = "<string with a description of the policy, including node and topic name>";
+auto policy_x_value = node->declare_parameter(
+    "<name of the parameter>", qos_provided_in_code.<getter_for_policy_x>(), policy_x_description);
+// Repeat for all policies
+rclcpp::QoS final_qos();
+// construct `final_qos` from the parameter values read above
+if (!user_provided_callback(final_qos) {
+    // qos profile rejected, throw error here
+}
+```
+
+In that way, we avoid installing a parameter callback that will be triggered when any parameter is modified.
+
+Note: this seems to be what you generally want in the case of read-only parameters.
+Note 2: We could also have a `declare_parameters_atomically` method and use:
+  ```cpp
+  auto callback_handle = node->add_on_set_parameters_callback(wrapper_for_qos_callback(user_provided_callback));
+  node->declare_parameters_atomically(...);  // names and defaults for all parameters given here
+  // callback_handle goes out of scope, the callback is auto-removed.
+  ```
+  `declare_parameters_atomically` would be needed, if not you cannot have conditions like `history == keep_all or history_depth > 10`.
```

Comment:
You should at least set the QoS parameters for a single topic atomically, but honestly, maybe all of them should be set atomically. What if you wanted `publisherA.history == subscriptionB.history`?

So I agree we need a `declare_parameters_atomically()`.

Reply:
> What if you wanted publisherA.history == subscriptionB.history?

That's a good question, but I think in that case the user can control in which order they create entities:

```cpp
publisher = node->create_publisher(
  "my_topic",
  my_pub_default_qos,
  [] (const QoSProfile & qos) -> bool {
      return qos_profile_is_valid(qos);
  });

subscription = node->create_subscription(
  "my_topic",
  my_sub_default_qos,
  subscription_callback
  [required_history=publisher->get_qos().get_history()] (const QoSProfile & qos) -> bool {
      return qos.get_history() == required_history && qos_profile_is_valid_2(qos);
  });
```

Declaring all qos parameters atomically sounds like a valid alternative, but I'm not sure when that should happen.